dialogue-core/src/main/java/com/palantir/dialogue/core/NodeSelectionStrategyChannel.java

```diff
@@ -0,0 +1,191 @@
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicReference;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.immutables.value.Value;
+
+public final class NodeSelectionStrategyChannel implements LimitedChannel {
+    private static final String NODE_SELECTION_HEADER = "Node-Selection-Strategy";
+
+    private final AtomicReference<ChannelWithStrategy> nodeSelectionStrategy;
+    private final AtomicReference<ImmutableList<LimitedChannel>> nodeChannels;
+    private final String channelName;
+    private final Random random;
+    private final Ticker tick;
+    private final TaggedMetricRegistry metrics;
+    private final NodeSelectionStrategySelector strategySelector;
+    private final LimitedChannel delegate;
+
+    public NodeSelectionStrategyChannel(
+            String channelName,
+            Random random,
+            Ticker tick,
+            TaggedMetricRegistry metrics,
+            NodeSelectionStrategySelector strategySelector) {
+        this.channelName = channelName;
+        this.random = random;
+        this.tick = tick;
+        this.metrics = metrics;
+        this.strategySelector = strategySelector;
+        this.nodeChannels = new AtomicReference<>(ImmutableList.of());
+        this.nodeSelectionStrategy = new AtomicReference<>(getUpdatedNodeSelectionStrategy(
+                null, nodeChannels.get(), strategySelector.getCurrentStrategy(), metrics, random, tick, channelName));
+        this.delegate = new SupplierChannel(() -> Preconditions.checkNotNull(
+                        nodeSelectionStrategy.get(), "node selection strategy must not be null")
+                .channel());
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        return delegate.maybeExecute(endpoint, request);
+    }
+
+    public void updateChannels(ImmutableList<LimitedChannel> newChannels) {
+        DialogueNodeSelectionStrategy updatedStrategy = strategySelector.setActiveChannels(newChannels);
+        nodeChannels.set(newChannels);
+        nodeSelectionStrategy.getAndUpdate(previousChannel -> getUpdatedNodeSelectionStrategy(
+                previousChannel.channel(), newChannels, updatedStrategy, metrics, random, tick, channelName));
+    }
+
+    private void updateRequestedStrategies(LimitedChannel channel, List<DialogueNodeSelectionStrategy> strategies) {
+        DialogueNodeSelectionStrategy updatedStrategy = strategySelector.updateChannelStrategy(channel, strategies);
+        nodeSelectionStrategy.getAndUpdate(currentStrategy -> {
```

Comment:
cas operations are relatively expensive, I think we want to do a quick check to avoid the cas in the expected case where the strategy doesn't change.